README.md
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=ptonini_pod-ca-injector&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=ptonini_pod-ca-injector)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=ptonini_pod-ca-injector&metric=coverage)](https://sonarcloud.io/summary/new_code?id=ptonini_pod-ca-injector)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=ptonini_pod-ca-injector&metric=bugs)](https://sonarcloud.io/summary/new_code?id=ptonini_pod-ca-injector)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=ptonini_pod-ca-injector&metric=security_rating)](https://sonarcloud.io/summary/new_code?id=ptonini_pod-ca-injector)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=ptonini_pod-ca-injector&metric=vulnerabilities)](https://sonarcloud.io/summary/new_code?id=ptonini_pod-ca-injector)

# Go API Server for kac

This is a generic definition for a Kubernetes Admission Controller. It implements both validate an mutate routes

## Overview
This server was generated by the [openapi-generator]
(https://openapi-generator.tech) project.
By using the [OpenAPI-Spec](https://github.com/OAI/OpenAPI-Specification) from a remote server, you can easily generate a server stub.
-

To see how to make this your own, look here:

[README](https://openapi-generator.tech)

- API version: 1.0.0
- Build date: 2022-07-26T01:04:19.810700-03:00[America/Fortaleza]

### Running the server

To run the server, follow these simple steps:

```
go run main.go
```

To run the server in a docker container
```
docker build --network=host -t kac .
```

Once the image is built, just run
```
docker run --rm -it kac 
```

### Known Issue

Endpoints sharing a common path may result in issues. For example, `/v2/pet/findByTags` and `/v2/pet/:petId` will result in an issue with the Gin framework. For more information about this known limitation, please refer to [gin-gonic/gin#388](https://github.com/gin-gonic/gin/issues/388) for more information.

A workaround is to manually update the path and handler. Please refer to [gin-gonic/gin/issues/205#issuecomment-296155497](https://github.com/gin-gonic/gin/issues/205#issuecomment-296155497) for more information.


CHANGELOG.md
# Changelog

## [1.2.1](https://github.com/ptonini/pod-ca-injector/compare/v1.2.0...v1.2.1) (2022-10-24)


### Bug Fixes

* added config.yaml to container image ([cf24a9c](https://github.com/ptonini/pod-ca-injector/commit/cf24a9ce8af65fa6dcb53bf4995ec304786da1bf))

## [1.2.0](https://github.com/ptonini/pod-ca-injector/compare/v1.1.0...v1.2.0) (2022-10-24)


### Features

* added environment variables configuration ([258f3d3](https://github.com/ptonini/pod-ca-injector/commit/258f3d3586b2ec266ea470fb659386dcd1912833))
* added environment variables configuration ([afaa83e](https://github.com/ptonini/pod-ca-injector/commit/afaa83e02e633aa6e704742ad79cac3f231d10d6))


### Bug Fixes

* corrected secret decoding ([c79ae47](https://github.com/ptonini/pod-ca-injector/commit/c79ae47330d354bd86a3e13e6c4160603ed310b2))

## [1.1.0](https://github.com/ptonini/pod-ca-injector/compare/v1.0.0...v1.1.0) (2022-10-21)


### Features

* added support for secrets and configmaps as rootca sources ([30a2eed](https://github.com/ptonini/pod-ca-injector/commit/30a2eedaecef1db54fdb480586385a16f08ed401))

## 1.0.0 (2022-10-19)


### Miscellaneous Chores

* first release ([ad2014b](https://github.com/ptonini/pod-ca-injector/commit/ad2014b501b07de56a93a6fae6bd52f6d9adc78b))


kac/config_test.go
package kac

import (
	"context"
	"encoding/base64"
	"fmt"
	"github.com/spf13/viper"
	"github.com/stretchr/testify/assert"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"os"
	"testing"
)

const (
	configFile      = "../config.yaml"
	testCertificate = `-----BEGIN CERTIFICATE-----
MIIDazCCAlOgAwIBAgIUIQDtYdVyYzIapvNndRrYojB+STUwDQYJKoZIhvcNAQEL
BQAwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgMClNvbWUtU3RhdGUxITAfBgNVBAoM
GEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZDAeFw0yMjEwMjMxOTA4MzVaFw0yMjEx
MjIxOTA4MzVaMEUxCzAJBgNVBAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRlMSEw
HwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0eSBMdGQwggEiMA0GCSqGSIb3DQEB
AQUAA4IBDwAwggEKAoIBAQC7+i18Z764qsA1abjYvXbO5GSm+xfmaLtjNV+mEf+Z
M7G6cwkGbLERVCeX9sv6eCERCX8loeQP/8WzHsU2npeKeOGhfDsILMexUS7NIRsg
KmzTdv05rfwXAiuUmRjobYf1RIvePV0RO1xv5F7h6ecUDWyhUmtpO+m+T5SSzqRY
AXakZpmhvx8QlKHrS66umDXWm0sR9cYRnAxO6SUtIoLQfYjmFNTpMy6sz3jxezfC
J9YfsEiAS2tH88Q72+kuczQDfz1O3XIKME68ExeWDmRlqeteOPSyGB72LmtBjOgM
zKSfPiD2wzqOfDBAyRhNzg8rNAb9bTMjzKyNOtlK8HAFAgMBAAGjUzBRMB0GA1Ud
DgQWBBT8I7ZHLIlHTN6DvVMCfp1MAVUiqDAfBgNVHSMEGDAWgBT8I7ZHLIlHTN6D
vVMCfp1MAVUiqDAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQCL
DZXT82lpcHMwW7lydHn7yx/h++kiYAUZkpOvm6hI2oeQo/OpdgMJIA0KZ/+cbyse
ataa2O6roLmXdv0CnAstVarlggewFPnuVFaBHj4iemH4jvDhpIvJCF62xrSdRpOk
jreVFmb9yvpFHTUInUUcQD/YCSz/dSBJv6ZEK4RpAwPsCwcEya2ijAfc1JpfbIz7
lwArCYp7dk24iqqvHyisEA74KzEIXgec4V443aEnVT0YFrPcCrKTyWO2B1cdg2iF
H9cXQE2NR3ZgAdEWuN5E9M5DA4NufOdEn106iEVloLNlGJ1JA8IyRwn4hRlX9RYa
5qgiE0YULCXL9x9SQDNN
-----END CERTIFICATE-----`
)

var (
	testRootCa = fmt.Sprintf(`{
  "test-url": {"type": "url", "source": "https://curl.se/ca/cacert.pem"}, 
  "test-local": {"type": "local", "source": %q }
}`, testCertificate)
	configMap = &corev1.ConfigMap{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "v1",
			Kind:       "ConfigMap",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name:      "test-config",
			Namespace: "default",
		},
		Data: map[string]string{
			"cert.crt": testCertificate,
		},
	}
	secret = &corev1.Secret{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "v1",
			Kind:       "Secret",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name:      "test-secret",
			Namespace: "default",
		},
		Data: map[string][]byte{
			"cert.crt": []byte(base64.StdEncoding.EncodeToString([]byte(testCertificate))),
		},
	}
)

func Test_Config(t *testing.T) {

	ctx := context.Background()
	ctx = context.WithValue(ctx, keyFakeClientSet, true)
	ctx = context.WithValue(ctx, keyFakeObjects, []runtime.Object{configMap, secret})

	_ = os.Setenv("CA_INJECTOR_ANNOTATIONS_INJECT", "ptonini.github.io/inject-ca")
	_ = os.Setenv("CA_INJECTOR_ANNOTATIONS_INJECTED", "ptonini.github.io/ca-injected")
	_ = os.Setenv("CA_INJECTOR_CONFIGMAP_NAME", "ca-injector")
	_ = os.Setenv("CA_INJECTOR_ROOTCA", testRootCa)

	t.Run("test read config", func(t *testing.T) {
		LoadConfig(configFile)
		_, err := getConfig()
		assert.NoError(t, err)
	})

	t.Run("test valid config from configmap", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "configMap", "source": "default/test-config/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.NoError(t, fetchBundles(ctx))
	})

	t.Run("test valid config from secret", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "secret", "source": "default/test-secret/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.NoError(t, fetchBundles(ctx))
	})

	t.Run("test valid config from configmap, no fake client", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "configMap", "source": "default/test-configmap/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(context.WithValue(ctx, keyFakeClientSet, false)))
	})

	t.Run("test valid config from secret, no fake client", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "secret", "source": "default/test-secret/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(context.WithValue(ctx, keyFakeClientSet, false)))
	})

	t.Run("test valid config from nonexistant secret", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "secret", "source": "default/fake/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(ctx))
	})

	t.Run("test valid config from nonexistant configMap", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "configMap", "source": "default/fake/cert.crt"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(ctx))
	})

	t.Run("test config with invalid bundle url", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "url", "source": "https://invalid.local"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(ctx))
	})

	t.Run("test config with invalid bundle content", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"remote": {"type": "url", "source": "https://example.com"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(ctx))
	})

	t.Run("test config with invalid local content", func(t *testing.T) {
		viper.Set("bundles", nil)
		_ = os.Setenv("CA_INJECTOR_ROOTCA", `{"local": {"type": "local", "source": "invalid"}}`)
		_ = readConfig("../config.yaml")
		assert.Error(t, fetchBundles(ctx))
	})

}


kac/routers_test.go
package kac

import (
	"context"
	"encoding/json"
	"github.com/gin-gonic/gin"
	"github.com/stretchr/testify/assert"
	admissionv1 "k8s.io/api/admission/v1"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"log"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
)

var (
	configMapsGVR = metav1.GroupVersionResource{
		Version:  "v1",
		Resource: "ConfigMaps",
	}
	pod = corev1.Pod{
		TypeMeta: metav1.TypeMeta{
			Kind:       "Pod",
			APIVersion: "v1",
		},
		ObjectMeta: metav1.ObjectMeta{Name: "test-pod"},
		Spec: corev1.PodSpec{
			Volumes: []corev1.Volume{},
			Containers: []corev1.Container{
				{
					VolumeMounts: []corev1.VolumeMount{},
				},
			},
		},
	}
)

func admissionReviewFactory(gvr metav1.GroupVersionResource, obj interface{}) string {
	rawObject, _ := json.Marshal(obj)
	a, err := json.Marshal(admissionv1.AdmissionReview{
		TypeMeta: metav1.TypeMeta{
			Kind:       "AdmissionReview",
			APIVersion: "admission.k8s.io/v1",
		},
		Request: &admissionv1.AdmissionRequest{
			Namespace: "default",
			Resource:  gvr,
			Object: runtime.RawExtension{
				Raw: rawObject,
			},
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	return string(a)
}

func fakeRequest(ctx context.Context, r *gin.Engine, method string, route string, rawBody string) *httptest.ResponseRecorder {
	w := httptest.NewRecorder()
	req, _ := http.NewRequest(method, route, strings.NewReader(rawBody))
	req = req.WithContext(ctx)
	r.ServeHTTP(w, req)
	return w
}

func Test_HealthcheckRoute(t *testing.T) {
	router := NewRouter()
	w := fakeRequest(context.Background(), router, http.MethodGet, "/health", "")
	assert.Equal(t, http.StatusOK, w.Code)
	assert.Equal(t, `{"status":"ok"}`, w.Body.String())
}

func Test_ReviewerRoutes(t *testing.T) {

	ctx := context.Background()
	ctx = context.WithValue(ctx, keyFakeObjects, []runtime.Object{configMap, secret})

	router := NewRouter()

	_ = os.Setenv("CA_INJECTOR_ANNOTATIONS_INJECT", "ptonini.github.io/inject-ca")
	_ = os.Setenv("CA_INJECTOR_ANNOTATIONS_INJECTED", "ptonini.github.io/ca-injected")
	_ = os.Setenv("CA_INJECTOR_CONFIGMAP_NAME", "ca-injector")
	_ = os.Setenv("CA_INJECTOR_ROOTCA", testRootCa)
	_ = readConfig("../config.yaml")
	_ = fetchBundles(ctx)
	config, _ := getConfig()

	for _, route := range []string{"/mutate", "/validate"} {
		t.Run("test route "+route+" with nil body", func(t *testing.T) {
			w := httptest.NewRecorder()
			req, _ := http.NewRequest(http.MethodPost, route, nil)
			router.ServeHTTP(w, req)
			assert.Equal(t, http.StatusBadRequest, w.Code)
		})
		t.Run("test route "+route+" with empty body", func(t *testing.T) {
			w := fakeRequest(ctx, router, http.MethodPost, route, "")
			assert.Equal(t, http.StatusBadRequest, w.Code)
		})
		t.Run("test route "+route+" with invalid body", func(t *testing.T) {
			invalidBody, _ := json.Marshal(configMap)
			w := fakeRequest(ctx, router, http.MethodPost, route, string(invalidBody))
			assert.Equal(t, http.StatusBadRequest, w.Code)
		})
	}

	t.Run("test route /validate with valid request", func(t *testing.T) {
		body := admissionReviewFactory(podsGVR, pod)
		w := fakeRequest(ctx, router, http.MethodPost, "/validate", body)
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("test route /mutate with invalid admission request resource", func(t *testing.T) {
		body := admissionReviewFactory(configMapsGVR, configMap)
		w := fakeRequest(ctx, router, http.MethodPost, "/mutate", body)
		assert.Equal(t, http.StatusInternalServerError, w.Code)
	})

	t.Run("test route /mutate with invalid admission request resource kind", func(t *testing.T) {
		body := admissionReviewFactory(podsGVR, configMap)
		w := fakeRequest(ctx, router, http.MethodPost, "/mutate", body)
		assert.Equal(t, http.StatusInternalServerError, w.Code)
	})

	t.Run("test route /mutate with valid request missing annotation", func(t *testing.T) {
		body := admissionReviewFactory(podsGVR, pod)
		w := fakeRequest(ctx, router, http.MethodPost, "/mutate", body)
		assert.Equal(t, http.StatusOK, w.Code)
	})

	pod.Annotations = map[string]string{config.Annotations.Inject: "test-url,test-local"}

	t.Run("test route /mutate with valid request, no fake client", func(t *testing.T) {
		body := admissionReviewFactory(podsGVR, pod)
		w := fakeRequest(ctx, router, http.MethodPost, "/mutate", body)
		assert.Equal(t, http.StatusInternalServerError, w.Code)
	})

	ctx = context.WithValue(ctx, keyFakeClientSet, true)

	t.Run("test route /mutate with valid request", func(t *testing.T) {
		body := admissionReviewFactory(podsGVR, pod)
		w := fakeRequest(ctx, router, http.MethodPost, "/mutate", body)
		assert.Equal(t, http.StatusOK, w.Code)
	})

}


main.go
/*
 * Kubernetes Admission Controller.
 * Copyright (C) 2022 Pedro Tonini
 * mailto:pedro DOT tonini AT hotmail DOT com
 *
 * Kubernetes Admission Controller is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Kubernetes Admission Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package main

import (
	"flag"
	"github.com/fsnotify/fsnotify"
	"github.com/spf13/viper"
	"log"

	"github.com/ptonini/pod-ca-injector/kac"
)

func main() {
	var tlsKey, tlsCert, configFile string
	flag.StringVar(&tlsKey, "tlsKey", "/certs/tls.key", "Path to the TLS key")
	flag.StringVar(&tlsCert, "tlsCert", "/certs/tls.crt", "Path to the TLS certificate")
	flag.StringVar(&configFile, "configFile", "/config/config.yaml", "Path to the TLS certificate")
	flag.Parse()

	kac.LoadConfig(configFile)
	viper.OnConfigChange(func(in fsnotify.Event) {
		kac.LoadConfig(configFile)
	})
	viper.WatchConfig()

	log.Printf("Server started")
	router := kac.NewRouter()
	log.Fatal(router.RunTLS(":8443", tlsCert, tlsKey))
}


kac/api.go
/*
 * Kubernetes Admission Controller.
 * Copyright (C) 2022 Pedro Tonini
 * mailto:pedro DOT tonini AT hotmail DOT com
 *
 * Kubernetes Admission Controller is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Kubernetes Admission Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package kac

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

func Health(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"status": "ok"})
}

// Mutate -
func Mutate(c *gin.Context) {
	serve(c, mutationReviewer)
}

// Validate -
func Validate(c *gin.Context) {
	serve(c, validationReviewer)
}


kac/helpers.go
/*
 * Kubernetes Admission Controller.
 * Copyright (C) 2022 Pedro Tonini
 * mailto:pedro DOT tonini AT hotmail DOT com
 *
 * Kubernetes Admission Controller is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Kubernetes Admission Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package kac

import (
	"context"
	"fmt"
	"io"
	"net/http"

	admissionv1 "k8s.io/api/admission/v1"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/apimachinery/pkg/runtime/serializer"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/kubernetes/fake"
	"k8s.io/client-go/rest"

	"github.com/gin-gonic/gin"
)

const (
	keyFakeClientSet = "fakeClientSet"
	keyFakeObjects   = "fakeObjects"
)

var (
	runtimeScheme = runtime.NewScheme()
	codecFactory  = serializer.NewCodecFactory(runtimeScheme)
	deserializer  = codecFactory.UniversalDeserializer()
)

type AdmissionReviewer func(context.Context, admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error)

func init() {
	_ = corev1.AddToScheme(runtimeScheme)
	_ = admissionv1.AddToScheme(runtimeScheme)
}

func errorResponse(c *gin.Context, statusCode int, err error) {
	_ = c.Error(err)
	c.JSON(statusCode, gin.H{"error": err.Error()})
}

func validateAndDeserialize(ar admissionv1.AdmissionReview, expectedGVR metav1.GroupVersionResource,
	expectedGVK schema.GroupVersionKind) (runtime.Object, error) {
	// Validate resource type
	if ar.Request.Resource != expectedGVR {
		return nil, fmt.Errorf("expect resource to be %s, got %s", expectedGVR, ar.Request.Resource)
	}
	// Deserialize object from AdmissionRequest
	obj, gvk, err := deserializer.Decode(ar.Request.Object.Raw, nil, nil)
	if *gvk != expectedGVK {
		return nil, fmt.Errorf("deserialized object is invalid: %v", obj)
	} else if err != nil {
		return nil, err
	} else {
		return obj, nil
	}
}

func getKubernetesClientSet(ctx context.Context) (kubernetes.Interface, error) {
	if ctx.Value(keyFakeClientSet) != nil && ctx.Value(keyFakeClientSet).(bool) {
		var objList []runtime.Object
		if ctx.Value(keyFakeObjects) != nil {
			objList = ctx.Value(keyFakeObjects).([]runtime.Object)
		}
		c := fake.NewSimpleClientset(objList...)
		return c, nil
	} else {
		config, err := rest.InClusterConfig()
		if err != nil {
			return nil, err
		}
		return kubernetes.NewForConfig(config)
	}

}

func serve(c *gin.Context, admissionReviewer AdmissionReviewer) {

	var resp *admissionv1.AdmissionReview
	var body []byte
	var ctx = c.Request.Context()

	if c.Request.Body != nil {
		body, _ = io.ReadAll(c.Request.Body)
	} else {
		errorResponse(c, http.StatusBadRequest, fmt.Errorf("request body is empty"))
		return
	}

	if obj, gvk, err := deserializer.Decode(body, nil, nil); err != nil {
		errorResponse(c, http.StatusBadRequest, err)
		return
	} else {
		req, ok := obj.(*admissionv1.AdmissionReview)
		if !ok {
			errorResponse(c, http.StatusBadRequest, fmt.Errorf("expected v1.AdmissionReview but got: %T", obj))
			return
		}
		resp = &admissionv1.AdmissionReview{}
		resp.SetGroupVersionKind(*gvk)
		resp.Response, err = admissionReviewer(ctx, *req)
		if err != nil {
			errorResponse(c, http.StatusInternalServerError, err)
			return
		}
		resp.Response.UID = req.Request.UID

	}

	c.JSON(http.StatusOK, resp)

}


kac/routers.go
/*
 * Kubernetes Admission Controller.
 * Copyright (C) 2022 Pedro Tonini
 * mailto:pedro DOT tonini AT hotmail DOT com
 *
 * Kubernetes Admission Controller is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Kubernetes Admission Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package kac

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// Route is the information for every URI.
type Route struct {
	// Name is the name of this Route.
	Name string
	// Method is the string for the HTTP method. ex) GET, POST etc..
	Method string
	// Pattern is the pattern of the URI.
	Pattern string
	// HandlerFunc is the handler function of this route.
	HandlerFunc gin.HandlerFunc
}

// Routes is the list of the generated Route.
type Routes []Route

// NewRouter returns a new router.
func NewRouter() *gin.Engine {
	gin.SetMode(gin.ReleaseMode)
	router := gin.Default()
	for _, route := range routes {
		switch route.Method {
		case http.MethodGet:
			router.GET(route.Pattern, route.HandlerFunc)
		case http.MethodPost:
			router.POST(route.Pattern, route.HandlerFunc)
		}
	}
	return router
}

var routes = Routes{
	{
		"Health",
		http.MethodGet,
		"/health",
		Health,
	},
	{
		"Mutate",
		http.MethodPost,
		"/mutate",
		Mutate,
	},
	{
		"Validate",
		http.MethodPost,
		"/validate",
		Validate,
	},
}


kac/config.go
/*
 * Kubernetes Admission Controller.
 * Copyright (C) 2022 Pedro Tonini
 * mailto:pedro DOT tonini AT hotmail DOT com
 *
 * Kubernetes Admission Controller is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Kubernetes Admission Controller is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package kac

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"github.com/spf13/viper"
	"io/ioutil"
	"k8s.io/apimachinery/pkg/apis/meta/v1"
	"log"
	"net/http"
	"path/filepath"
	"strings"
)

type Config struct {
	Annotations struct {
		Inject   string
		Injected string
	}
	ConfigMapName string
	RootCA        map[string]struct {
		Type   string
		Source string
	}
	Bundles map[string]string
}

func LoadConfig(configFile string) {
	log.Printf("Loding config")
	err := readConfig(configFile)
	if err != nil {
		log.Fatal(err)
	}
	err = fetchBundles(context.Background())
	if err != nil {
		log.Fatal(err)
	}
}

func readConfig(configFile string) error {
	viper.SetConfigName(strings.Split(filepath.Base(configFile), ".")[0])
	viper.AddConfigPath(filepath.Dir(configFile))
	viper.MustBindEnv("configmapname", "CA_INJECTOR_CONFIGMAP_NAME")
	viper.MustBindEnv("annotations.inject", "CA_INJECTOR_ANNOTATIONS_INJECT")
	viper.MustBindEnv("annotations.injected", "CA_INJECTOR_ANNOTATIONS_INJECTED")
	viper.MustBindEnv("rootca", "CA_INJECTOR_ROOTCA")
	return viper.ReadInConfig()
}

func fetchBundles(ctx context.Context) error {
	config, err := getConfig()
	for k, v := range config.RootCA {
		var bundle string
		log.Printf("Fetching bundle %s from %s: %q", k, v.Type, v.Source)
		switch v.Type {
		case "local":
			if err = validateCertificate(v.Source); err == nil {
				bundle = v.Source
			}
		case "url":
			bundle, err = getBundleFromURL(v.Source)
		case "secret":
			s := strings.Split(v.Source, "/")
			bundle, err = getCAFromSecret(ctx, s[0], s[1], s[2])
		case "configMap":
			s := strings.Split(v.Source, "/")
			bundle, err = getCAFromConfigMap(ctx, s[0], s[1], s[2])
		}
		if err != nil {
			return err
		}
		viper.Set(fmt.Sprintf(`bundles.%s`, k), bundle)
	}
	return err
}

func getConfig() (*Config, error) {
	var config Config
	err := viper.Unmarshal(&config)
	if err != nil {
		err = json.Unmarshal([]byte(viper.Get("rootca").(string)), &config.RootCA)
	}
	return &config, err
}

func getBundleFromURL(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	body, _ := ioutil.ReadAll(resp.Body)
	defer func() { _ = resp.Body.Close() }()
	bodyStr := string(body)
	return bodyStr, validateCertificate(bodyStr)
}

func getCAFromSecret(ctx context.Context, ns string, secretName string, key string) (string, error) {
	clientSet, err := getKubernetesClientSet(ctx)
	if err != nil {
		return "", err
	}
	secret, err := clientSet.CoreV1().Secrets(ns).Get(ctx, secretName, v1.GetOptions{})
	if err != nil {
		return "", err
	}
	certificateBytes, _ := base64.StdEncoding.DecodeString(string(secret.Data[key]))
	certificate := string(certificateBytes)
	return certificate, validateCertificate(certificate)
}

func getCAFromConfigMap(ctx context.Context, ns string, configMapName string, key string) (string, error) {
	clientSet, err := getKubernetesClientSet(ctx)
	if err != nil {
		return "", err
	}
	configMap, err := clientSet.CoreV1().ConfigMaps(ns).Get(ctx, configMapName, v1.GetOptions{})
	if err != nil {
		return "", err
	}
	certificate := configMap.Data[key]
	return certificate, validateCertificate(certificate)
}

func validateCertificate(bundle string) error {
	block, _ := pem.Decode([]byte(bundle))
	if block == nil || block.Type != "CERTIFICATE" {
		return fmt.Errorf("invalid certificate")
	}
	return nil
}
